runtime/src/main/java/org/corfudb/runtime/object/MultiVersionObject.java

```diff
@@ -0,0 +1,303 @@
+        this.lock = new StampedLock();
+
+        this.mvoCache.registerMVO(this.streamID,this);
+    }
+
+    /**
+     * Get snapshot proxy for a given timestamp.
+     * 1. If called from non-transactional access, timestamp is stream tail and
+     *    there should be an exact match in MVOCache if not yet evicted.
+     * 2. If called from transactional access, timestamp is transaction snapshot
+     *    address
+     *    a. if timestamp > highestResolvedAddress, query sequencer for latest updates.
+     *    b. if timestamp <= highestResolvedAddress, run MVOCache::floorEntry()
+     *
+     * @param timestamp the timestamp to query
+     * @return the snapshot proxy against which the snapshot access is run
+     */
+    public ICorfuSMRSnapshotProxy<T> getSnapshotProxy(long timestamp) {
+
+        // 1. Exact match
+        Optional<ICorfuSMRSnapshotProxy<T>> snapshotProxyOptional = mvoCache.get(
+                new VersionedObjectIdentifier(streamID, timestamp),
+                this::getSMRSnapshotProxy
+        );
+        if (snapshotProxyOptional.isPresent()) {
+            return snapshotProxyOptional.get();
+        }
+
+        // 2. Transaction accesses
+        if (timestamp > highestResolvedAddress || timestamp == Address.NON_EXIST) {
+            // 2.a need to query sequencer for latest updates
+            AtomicLong vloAccessedVersion = new AtomicLong();
+            snapshotProxyOptional = getVersionedObjectUnderLock(timestamp, vloAccessedVersion::set);
+
+            if (snapshotProxyOptional.isPresent()) {
+                correctnessLogger.trace("Version, {}", vloAccessedVersion.get());
+                log.trace("Access [{}] Updated (writelock) access at {}", this, vloAccessedVersion.get());
+                return snapshotProxyOptional.get();
+            }
+        } else {
+            // 2.b address has already been resolved
+            snapshotProxyOptional = mvoCache.floorEntry(
+                    new VersionedObjectIdentifier(streamID, timestamp),
+                    this::getSMRSnapshotProxy
+            );
+            if (snapshotProxyOptional.isPresent()) {
+                return snapshotProxyOptional.get();
+            }
+        }
+
+        throw new StaleObjectVersionException(streamID, timestamp);
+    }
+
+    protected Optional<ICorfuSMRSnapshotProxy<T>> getVersionedObjectUnderLock(long timestamp, Consumer<Long> versionAccessed) {
+        // polling is expensive
+        while (timestamp <= currentSyncVersion && lock.tryWriteLock() == 0) {
+            Optional<ICorfuSMRSnapshotProxy<T>> snapshotProxyOptional = mvoCache.get(
+                    new VersionedObjectIdentifier(streamID, timestamp), snapshotReferenceProxyGenerator);
+            if (snapshotProxyOptional.isPresent()) {
+                return snapshotProxyOptional;
+            }
+        }
+
+        long ts = 0;
+        try {
+            ts = lock.writeLock();
+            currentSyncVersion = timestamp;
```

Comment:
This looks like a race. Other threads will observe that `timestamp <= currentSyncVersion` but before the sync happens (i.e., snapshotProxyOptional.isPresent() is false).